docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,277 @@
+# Kubelet and systemd interaction
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+This document presents a specification for how the `kubelet` interfaces with `systemd`.
+
+## Motivation
+
+Many Linux distributions have either adopted, or plan to adopt `systemd` as their init system.
+
+This document describes a series of enhancements that should be made to the `kubelet` to better 
+integrate with these distributions independent of their chosen container runtime.
+
+## Scope of proposal
+
+This proposal does not account for running the `kubelet` in a container.
+
+## systemd units
+
+To help understand this proposal, we first provide a brief summary of `systemd` behavior.
+
+`systemd` manages a hierarchy of `slice`, `scope`, and `service` units.
+
+* `service` - application on the server that is launched by `systemd`; how it should start/stop; 
+when it should be started; under what circumstances it should be restarted; and any resource 
+controls that should be applied to it.
+* `scope` - a process or group of processes which are not launched by `systemd` (i.e. fork), like
+a service, resource controls may be applied
+* `slice` - organizes a hierarchy in which `scope` and `service` units are placed.  a `slice` may
+contain `slice`, `scope`, or `service` units; processes are attached to `service` and `scope`
+units only, not to `slices`. The hierarchy is intended to be unified, meaning a process may
+only belong to a single leaf node.
+
+## cgroup hierarchy: orthogonal versus unified hierarchies and cgroupfs single writer
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process could
+exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not
+always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated
+separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy
+across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the
+per-controller hierarchies are eliminated in favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process may only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than
+allowing multiple processes to independently write to parts of the file-system.
```

Comment:
Suggestion to add:

> Systemd is in general the single-writer on all cgroups but services and scopes can be configured with [the `Delegate=` option](http://man7.org/linux/man-pages/man5/systemd.resource-control.5.html#OPTIONS). In this case, systemd knows that another writer will handle this part of the cgroup tree.

I think rkt does not explain this correctly. I filed a bug in rkt for this: https://github.com/coreos/rkt/issues/1883